src/planning.py

```diff
@@ -264,46 +288,52 @@ def _run_low_level_search(task: Task, option_model: _OptionModelBase,
         if option.initiable(state):
             try:
                 next_state = option_model.get_next_state(state, option)
-                discovered_failures[cur_idx] = None  # no failure occurred
             except EnvironmentFailure as e:
                 can_continue_on = False
-                failure = _DiscoveredFailure(e, nsrt)
                 # Remember only the most recent failure.
-                discovered_failures[cur_idx] = failure
-                # If we're immediately propagating failures, raise it now.
-                if CFG.sesame_propagate_failures == "immediately":
-                    raise _DiscoveredFailureException("Discovered a failure",
-                                                      failure)
-            if not discovered_failures[cur_idx]:
+                discovered_failures[cur_idx] = _DiscoveredFailure(e, nsrt)
+                cur_idx += 1  # it's about to be decremented again
+            else:  # an EnvironmentFailure was not raised
+                discovered_failures[cur_idx] = None
                 traj[cur_idx + 1] = next_state
                 cur_idx += 1
                 # Check atoms against expected atoms_sequence constraint.
                 assert len(traj) == len(atoms_sequence)
                 # The expected atoms are ones that we definitely expect to be
-                # true at this point in the plan. They are not *all* the atoms
-                # that could be true.
+                # true at this point in the refinement. They are not *all* the
+                # atoms that could be true.
                 expected_atoms = {
                     atom
                     for atom in atoms_sequence[cur_idx]
                     if atom.predicate.name != _NOT_CAUSES_FAILURE
                 }
-                # This is equivalent to, but faster than, checking whether
-                # expected_atoms is a subset of utils.abstract(traj[cur_idx],
-                # predicates).
+                # This "if all" statement is equivalent to, but faster than,
+                # checking whether expected_atoms is a subset of
+                # utils.abstract(traj[cur_idx], predicates).
                 if all(atom.holds(traj[cur_idx]) for atom in expected_atoms):
                     can_continue_on = True
-                    if cur_idx == len(skeleton):  # success!
-                        result = plan
-                        return result
+                    if cur_idx == len(skeleton):
+                        return plan, True  # success!
                 else:
                     can_continue_on = False
-            else:
-                cur_idx += 1  # it's about to be decremented again
         else:
             # If the option is not initiable, need to resample / backtrack.
             can_continue_on = False
             cur_idx += 1  # it's about to be decremented again
         if not can_continue_on:
+            # Update the longest_failed_refinement found so far.
```

Comment:
@tomsilver i think this is the right place to do this, finally...

Reply:
note that this will only update longest_failed_refinement if we're at a step where we failed (for any reason: non-initiable option, expected_atoms check failed, or an EnvironmentFailure occurred). i think this is alright & exactly what we want -- that's what makes it a "failed" refinement

Reply:
yeah, this looks good